# JSON Data Wrangling Mini Project
***
---- Lily Elizabeth John
***
Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
+ Find the 10 countries with most projects
+ Find the top 10 major project themes (using column 'mjtheme_namecode')
+ You will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.
***

In [1]:
#import libraries
import numpy as np
import pandas as pd
import json
from pandas.io.json import json_normalize

In [2]:
#Read in json data using pandas read_json to flatten one hierarchy of objects in the json nested dictionary and describe data
projects_df=pd.read_json('data/world_bank_projects.json')
projects_df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Name': 'Education for all', 'Percent': 100}   
1  {'Name': 'Other economic management', 'Percent...   
2    {'Name': 'Regional integration', 'Percent': 46}   
3  {'Name': 'Participation and civic engagement',...   
4  {'Name': 'Export development and competitivene...   

                                      theme_namecode    themecode   totalamt  \
0      [{'name': 'Education for all', 'code': '65'}]           65  130000000   
1  [{'name': 'Other economic management', 'code':...        54,24          0   
2  [{'name': 'Regional integration', 'code': '47'...  52,81,25,47    6060000   
3  [{'name': 'Participation and civic engagement'...        59,57          0   
4  [{'name': 'Export development and competitiven...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [3]:
projects_df.columns

Index(['_id', 'approvalfy', 'board_approval_month', 'boardapprovaldate',
       'borrower', 'closingdate', 'country_namecode', 'countrycode',
       'countryname', 'countryshortname', 'docty', 'envassesmentcategorycode',
       'grantamt', 'ibrdcommamt', 'id', 'idacommamt', 'impagency',
       'lendinginstr', 'lendinginstrtype', 'lendprojectcost',
       'majorsector_percent', 'mjsector_namecode', 'mjtheme',
       'mjtheme_namecode', 'mjthemecode', 'prodline', 'prodlinetext',
       'productlinetype', 'project_abstract', 'project_name', 'projectdocs',
       'projectfinancialtype', 'projectstatusdisplay', 'regionname', 'sector',
       'sector1', 'sector2', 'sector3', 'sector4', 'sector_namecode',
       'sectorcode', 'source', 'status', 'supplementprojectflg', 'theme1',
       'theme_namecode', 'themecode', 'totalamt', 'totalcommamt', 'url'],
      dtype='object')

In [4]:
projects_df.shape

(500, 50)

### 1. Find the 10 countries with most projects

In [5]:
#Using groupby to count number of projects for each country and then sort in descending order to get the 10 countries with most projects
projects_df.groupby(['countryname']).size().sort_values(ascending=False).head(10)

countryname
People's Republic of China         19
Republic of Indonesia              19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Nepal                              12
People's Republic of Bangladesh    12
Kingdom of Morocco                 12
Africa                             11
Republic of Mozambique             11
dtype: int64

### 2. Find the top 10 major project themes (using column 'mjtheme_namecode')

In [6]:
#Review 'mjtheme_namecode'
projects_df['mjtheme_namecode'].head()

0    [{'name': 'Human development', 'code': '8'}, {...
1    [{'name': 'Economic management', 'code': '1'},...
2    [{'name': 'Trade and integration', 'code': '5'...
3    [{'name': 'Social dev/gender/inclusion', 'code...
4    [{'name': 'Trade and integration', 'code': '5'...
Name: mjtheme_namecode, dtype: object

In [7]:
#Load json and normalize json string to flatten the dictionary in field 'mjtheme_namecode'
data_json=json.load((open('data/world_bank_projects.json')))
mjtheme_df=json_normalize(data_json,'mjtheme_namecode',['id'])
#Get shape to verify the shape with final dataframe in question 3
mjtheme_df.shape

(1499, 3)

In [8]:
#Get count of project themes and sort in descending order in order to get top 10 major project themes
mjtheme_df.name.value_counts().sort_values(ascending=False).head(10)

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

Notes

+ 122 namecodes do not have the corresponding project theme names. 

In [9]:
#How many codes are there in each project
mjtheme_df.groupby(['id']).size().sort_values(ascending=False).head()

id
P129652    5
P127723    5
P128891    5
P128832    5
P128284    5
dtype: int64

Notes
+ There are a lot of projects associated with more than 1 project theme name.

In [10]:
#Let's explore a few projects associated with multiple project theme names.
print(mjtheme_df[mjtheme_df.id=='P129652'])

    code                         name       id
545    7  Social dev/gender/inclusion  P129652
546    2     Public sector governance  P129652
547    8            Human development  P129652
548   10            Rural development  P129652
549    8            Human development  P129652


Notes
+ Project P129652 has been linked to 'Human Development' twice.

In [11]:
#Are there many projects with duplicated associations
mjtheme_df.groupby(['id','code']).size().sort_values(ascending=False).head()

id       code
P129663  8       5
P130873  10      5
P131945  8       5
P131194  8       5
P131919  8       5
dtype: int64

Notes
+ Above listed projects are 5 projects among many that have duplicated associations. 
+ For instance: P129663 is associated with Human Development 5 times. This overestimates the project theme name count.

### 3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [12]:
#Replace missing values with np.nan to be able to work with dataset
mjtheme_df['name']=mjtheme_df['name'].replace('',np.nan)
#Subset dataset to exclude id
mjtheme_df1=mjtheme_df[['code','name']]
#Get unique combination of code and name from dataset and exclude NaNs
mjtheme_df2=mjtheme_df1.drop_duplicates()
mjtheme_lookuptable=mjtheme_df2[mjtheme_df2['name'].notnull()]
#Get only codes from initial dataframe and merge in names with code as key to get all values including missing
mjtheme_df1=pd.DataFrame(mjtheme_df1['code'])
mergedmjtheme_df=mjtheme_df1.merge(mjtheme_lookuptable,on='code',how='left')
#Verify that the shape matches original dataset
print(mergedmjtheme_df.shape)

(1499, 2)


In [13]:
#Get new count of project themes with filled in values
mj_with_dupes=pd.DataFrame(mergedmjtheme_df.name.value_counts().sort_values(ascending=False))
mj_with_dupes.columns=['with dupes']
mj_with_dupes=mj_with_dupes.reset_index()
mj_with_dupes.head(10)

index  with dupes
0  Environment and natural resources management         250
1                             Rural development         216
2                             Human development         210
3                      Public sector governance         199
4         Social protection and risk management         168
5      Financial and private sector development         146
6                   Social dev/gender/inclusion         130
7                         Trade and integration          77
8                             Urban development          50
9                           Economic management          38

Notes
+ The missing project themes have been replaced with accurate project theme names.


### 3 Extension: Redo counts after removing duplicate id, code combinations

In [14]:
#Subset dataset to exclude name and drop duplicate combinations of id and code
mjtheme_df_unique=mjtheme_df[['id','code']].drop_duplicates()
#Get only codes from deduped dataframe and merge in names with code as key to get all values including missing
mjtheme_df1_unique=pd.DataFrame(mjtheme_df_unique['code'])
mergedmjtheme_df_unique=mjtheme_df1_unique.merge(mjtheme_lookuptable,on='code',how='left')
#Verify that the shape matches original dataset
print(mergedmjtheme_df_unique.shape)

(1070, 2)


In [15]:
#Get new count of project themes with filled in values
mj_without_dupes=pd.DataFrame(mergedmjtheme_df_unique.name.value_counts().sort_values(ascending=False))
mj_without_dupes.columns=['without dupes']
mj_without_dupes=mj_without_dupes.reset_index()
mj_without_dupes.head(10)

index  without dupes
0  Environment and natural resources management            157
1                             Rural development            148
2                      Public sector governance            140
3                             Human development            128
4      Financial and private sector development            119
5         Social protection and risk management            116
6                   Social dev/gender/inclusion            114
7                         Trade and integration             61
8                             Urban development             40
9                           Economic management             33

In [16]:
#Combine both counts to compare and contrast
mj_count_diff=mj_with_dupes.merge(mj_without_dupes,how='outer')
mj_count_diff['count_diff']=mj_count_diff['with dupes']-mj_count_diff['without dupes']
mj_count_diff

index  with dupes  without dupes  \
0   Environment and natural resources management         250            157   
1                              Rural development         216            148   
2                              Human development         210            128   
3                       Public sector governance         199            140   
4          Social protection and risk management         168            116   
5       Financial and private sector development         146            119   
6                    Social dev/gender/inclusion         130            114   
7                          Trade and integration          77             61   
8                              Urban development          50             40   
9                            Economic management          38             33   
10                                   Rule of law          15             14   

    count_diff  
0           93  
1           68  
2           82  
3           59  
4           52  
5           27  
6           16  
7           16  
8           10  
9            5  
10           1

Notes:
+ After removing duplicate associations, 'Human Development' theme comes down in ranking to 4th.
+ Financial and Private Sector Development climbs higher in ranking to 5th spot moving Sc=ocial Protection and Risk Management 1 level down.